tests/integration/test_excludes.py

```diff
@@ -145,3 +160,204 @@ def test_requirements_pex_exclude(
+                    continue
+                yield cls(pip_version=pip_version, resolver_version=resolver_version)
+
+    pip_version = attr.ib()  # type: PipVersionValue
+    resolver_version = attr.ib()  # type: ResolverVersion.Value
+
+    def iter_args(self):
+        # type: () -> Iterator[str]
+        yield "--pip-version"
+        yield str(self.pip_version)
+        yield "--resolver-version"
+        yield str(self.resolver_version)
+
+    def __str__(self):
+        # type: () -> str
+        return "-".join(
+            (
+                str(self.pip_version),
+                "legacy" if self.resolver_version is ResolverVersion.PIP_LEGACY else "resolvelib",
+            )
+        )
+
+
+@pytest.mark.parametrize(
+    "pip_options",
+    [pytest.param(pip_options, id=str(pip_options)) for pip_options in PipOptions.iter()],
+)
+def test_exclude_deep(
+    tmpdir,  # type: Any
+    pip_options,  # type: PipOptions
+):
+    # type: (...) -> None
+
+    # Bootstrap the Pip version being used if needed before we turn off PyPI.
+    if pip_options.pip_version is not PipVersion.VENDORED:
+        run_pex_command(
+            args=list(pip_options.iter_args()) + ["ansicolors==1.1.8", "--", "-c", ""]
+        ).assert_success()
+
+    venv = Virtualenv.create(os.path.join(str(tmpdir), "venv"))
+    pip = venv.install_pip(upgrade=True)
+    subprocess.check_call(args=[pip, "install", "-U", "setuptools", "wheel"])
+
+    find_links = os.path.join(str(tmpdir), "find_links")
+    project_dir = os.path.join(str(tmpdir), "projects")
+
+    foo_dir = os.path.join(project_dir, "foo")
+    with safe_open(os.path.join(foo_dir, "foo.py"), "w") as fp:
+        fp.write(
+            dedent(
+                """\
+                from bar import BAR
+
+
+                def foo():
+                    return BAR * 42
+                """
+            )
+        )
+    with safe_open(os.path.join(foo_dir, "setup.py"), "w") as fp:
+        fp.write(
+            dedent(
+                """\
+                from setuptools import setup
+
+
+                setup(
+                    name="foo",
+                    version="0.1.0",
+                    install_requires=["bar"],
+                    py_modules=["foo"],
+                )
+                """
+            )
+        )
+    venv.interpreter.execute(
+        args=["setup.py", "bdist_wheel", "--dist-dir", find_links], cwd=foo_dir
+    )
+
+    bar_dir = os.path.join(project_dir, "bar")
+    with safe_open(os.path.join(bar_dir, "bar.py"), "w") as fp:
+        print("BAR=1", file=fp)
+    with safe_open(os.path.join(bar_dir, "setup.py"), "w") as fp:
+        fp.write(
+            dedent(
+                """\
+                import os
+                import sys
+
+                from setuptools import setup
+
+
+                if "BEHAVE" not in os.environ:
+                    sys.exit("I'm an evil package.")
```

Comment:
🤣

This is so true. Packages that put `sys.exit` in `setup.py` are EVIL!

Reply:
Well, I felt a bit bad about this. It's really Python that is evil here. The pyinotify project is ~17 years old and, IIRC there was no way at that time (just like today!) for an sdist to do anything but fail when you try to compile it on a platform it doesn't support. The sys.exit in the position pyinotify puts it is aggressive, but with out it the build will just fail later anyhow.

The real problem then, is the distutils / setuptools build system here. Since those require evaluating Python code (in greater degrees further back in time) to learn just project metadata, aggressive tends towards evil - the project sdist acts as a hand grenade for the only curious.

Setuptools grew setup.cfg eventually to become up to ~100% declarative and then later Python people (but crucially not Python!) woke up and forked PyPA and eventually produced PEP 517/518 and now setuptools even supports pyproject.toml declarative metadata.

I can't describe the brokenness of a language not having its story straight better than Indy Greg or the related discussion, or, the icing on the cake, the ensuing inaction:
+ https://gregoryszorc.com/blog/2023/10/30/my-user-experience-porting-off-setup.py/
+ https://discuss.python.org/t/user-experience-with-porting-off-setup-py/37502/1

It's true it can be hard for an OSS project to muster the resources to tackle a big project, but Python has some serious financial backers and it could decide to make this a priority and get the money and people-time to clean this up in, say, a 5 year period. Instead though its floundering and leaving a confusing landscape in its wake for users.